src/state.rs

```diff
@@ -227,12 +227,14 @@ impl State {
         });
     }
 
-    fn unregister_connection(&mut self, connection: Connection) {
-        if self.connections.remove(&connection).is_some() {
+    pub fn unregister_connection(&mut self, connection: Connection) {
+        if self.connections.remove(&connection).is_none() {
             // Only send the event if the connection was there
```

Comment:
I like the approach to return early if some condition prevent the logic. I saw Linus Torvalds defending this style too (a shame I didn't save the link).

Anyway... this comment looks half out of place once you moved the notification code. Maybe something like `// Avoid duplicate connection cleanup`?